klippy/extras/filament_switch_sensor.py

```diff
@@ -0,0 +1,165 @@
+            pause_resume.send_pause_command(from_event=True)
+        self._exec_gcode(self.runout_gcode)
+        self.event_running = False
+    def _autoload_event_handler(self, eventtime):
+        if self.event_running:
+            return
+        self.event_running = True
+        self._exec_gcode(self.autoload_gcode)
+        self.event_running = False
+    def _exec_gcode(self, script):
+        try:
+            self.gcode.run_script(script)
+        except self.gcode.error as err:
+            # We arent inside a gcode handler, so respond with an error
+            self.gcode.respond_error(str(err))
+        except:
+            raise
+    def set_callbacks(self, runout_cb=None, detected_cb=None, monitor=True):
+        # Allow other modules to receive sensor status notifications
+        if runout_cb is None:
+            runout_cb = self._runout_event_handler
+        if detected_cb is None:
+            detected_cb = self._autoload_event_handler
+        # When monitor is set to true, the filament_sensor module will monitor
+        # printer status.  If Klipper is printing, runout detection is enabled
+        # if the runout_gcode is set.  If Klipper is ready/idle, insert
+        # detection is enabled if the autoload_gcode is set.
+        self.monitor_state = monitor
+        self.runout_cb = runout_cb
+        self.insert_cb = detected_cb
+    def set_enable(self, runout, insert):
+        if runout and insert:
+            # both cannot be enabled
+            insert = False
+        self.runout_enabled = runout
+        self.insert_enabled = insert
+    def query_status(self):
+        raise NotImplementedError("Sensor must implement query_status")
+    cmd_QUERY_FILAMENT_SENSOR_help = "Query the status of the Filament Sensor"
+    def cmd_QUERY_FILAMENT_SENSOR(self, params):
+        msg = self.query_status()
+        self.gcode.respond_info(msg)
+
+SETTLE_TIME = .1
+
+class SwitchSensor(BaseSensor):
+    def __init__(self, config):
+        super(SwitchSensor, self).__init__(config)
+        self.printer = config.get_printer()
+        self.reactor = self.printer.get_reactor()
+        self.buttons = self.printer.try_load_module(config, 'buttons')
+        switch_pin = config.get('switch_pin')
+        self.buttons.register_buttons([switch_pin], self._button_handler)
+        self.event_delay = config.getfloat('event_delay', 3., above=0.)
+        self.event_button_state = None
+        self.last_button_state = False
+        self.last_cb_event_time = 0.
+        self.switch_event_timer = self.reactor.register_timer(
+            self._switch_event)
+    def _button_handler(self, eventtime, state):
+        self.last_button_state = state
+        if self.event_button_state is not None:
+            # currently waiting on for timer to execute
+            return
+        self.event_button_state = state
+        self.reactor.update_timer(
+            self.switch_event_timer, eventtime + SETTLE_TIME)
```

Comment:
I'm not sure what this is attempting to do, but it doesn't seem right. Timing in the host code is always a bit iffy because there could be variable amounts of delays between the events and their reception in the host. Better to get the actual event time (a clock/print_time from the micro-controller) or to use the sent/receive_time from the actual timestamp of the received micro-controller messages.

At a high-level I don't think the filament sensor should need to use the reactor at all.  The reactor is more intended to schedule system events - at first glance the filament sensor is just reacting to internal events.

Reply:
This timer makes sure that a change in switch state is held for ~100ms before executing a callback.   I see where the issue is here, I shouldn't use the eventtime to update the timer.

I moved this functionality to buttons.py.  There I use reactor.monotonic() to set the wake time.   This probably isn't absolutely necessary functionality, I thought it would be wise given that the switch will likely move around and have filament running past it.  Depending on the way the switch is mounted this could result in the switch triggering prematurely.  If you do not like it I have no problem removing it.

Reply:
FWIW, I'd say add support for extended debouncing when we get confirmation that it is required.  There's a good chance the troublesome hardware will need something different than anticipated.

Reply:
I'm battling with the cheap filament runout sensor that came with my Two Trees Sapphire Pro. It keeps giving false alerts and pausing my prints. I have a feeling a longer debounce could be helpful in my case